pytests/tests/test_pyclasses.py

```diff
@@ -25,3 +27,15 @@ def test_iter():
     with pytest.raises(StopIteration) as excinfo:
         next(i)
     assert excinfo.value.value == "Ended"
+
+
+def test_del():
+    d = pyclasses.PyClassDel()
+    del d
+
+
+@mock.patch("sys.unraisablehook")
+def test_del_error(unraisablehook):
+    d = pyclasses.PyClassDelError()
+    del d
+    assert unraisablehook.called
```

Comment:
This test case does not pass locally, adding a `panic!()` to `__del__` seems to indicate that `__del__` is not called as it did not trigger a panic.

Reply:
As per https://github.com/PyO3/pyo3/pull/2484#discussion_r907034479 this indicates to me that we may want to consider adding a block similar to the following to `PyCell::tp_dealloc`

https://github.com/python/cpython/blob/56310136170ef6653cb050f58dd2d32538997f59/Modules/_asynciomodule.c#L1534-L1542

But `PyObject_CallFinalizerFromDealloc` is only in the limited API, so it might be that we can only support `__del__` when not running `abi3`.

Reply:
That's unfortunate.

> ...or maybe we should be removing our custom tp_dealloc and moving all logic to tp_finalize ??

Could we do this instead so it does not require calling `PyObject_CallFinalizerFromDealloc` in extension module?

Reply:
Yes, that seems reasonable to me. I have no idea whether it will definitely work. I plan to play around with this once the 0.17 release is complete.